sklearn/feature_selection/from_model.py

```diff
@@ -183,10 +202,16 @@ class SelectFromModel(BaseEstimator, SelectorMixin):
     `threshold_`: float
         The threshold value used for feature selection.
     """
-    def __init__(self, estimator, threshold=None, prefit=False):
+    def __init__(self, estimator, threshold=None, prefit=False, k=None):
         self.estimator = estimator
         self.threshold = threshold
         self.prefit = prefit
+        self.k = k
+
+    def _check_params(self, X, y):
+        if self.k is not None and not (self.k == "all" or 0 <= self.k <= X.shape[1]):
```

Comment:
`X` has no shape attribute if it isn't a numpy array. Use `check_X_y` or `check_array` from `sklearn.utils.validation`.